# Importação das bibliotecas

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import dask.dataframe as dd
import matplotlib.pyplot as plt
import tensorflow as tf
from tensorflow.keras.models import Sequential
from keras.layers import Dense, Dropout, BatchNormalization, Conv1D, Flatten
from tensorflow.keras.optimizers import Adam
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from keras.callbacks import EarlyStopping
import matplotlib.pyplot as plt
from sklearn.metrics import roc_curve, auc
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dens
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, roc_auc_score
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import Adam
from sklearn.metrics import accuracy_score, recall_score, f1_score, roc_auc_score
from tensorflow.keras.optimizers import Adam


/usr/local/lib/python3.10/dist-packages/dask/dataframe/__init__.py:42: FutureWarning: 
Dask dataframe query planning is disabled because dask-expr is not installed.

You can install it with `pip install dask[dataframe]` or `conda install dask`.
This will raise in a future version.

  warnings.warn(msg, FutureWarning)


# Carregamento do aqruivo: cartão de crédito

In [10]:
# Carregar os dados
data = pd.read_csv('/content/drive/MyDrive/ponderada/creditcard (1).csv')



# Divisão conjunto treino e teste / normalização

In [20]:
X = data.drop('Class', axis=1)
y = data['Class']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Modelo 1 - o meu modelo original em que estarei utilizando o Keras

In [12]:
model = Sequential([
    Dense(64, input_dim=X_train.shape[1], activation='relu'),
    BatchNormalization(),
    Dropout(0.5),
    Dense(32, activation='relu'),
    BatchNormalization(),
    Dropout(0.5),
    Dense(1, activation='sigmoid')
])

model.compile(optimizer=Adam(learning_rate=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(X_train, y_train,
                    epochs=10,
                    batch_size=64,
                    validation_data=(X_test, y_test))

Epoch 1/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 20s 5ms/step - accuracy: 0.9126 - loss: 0.2101 - val_accuracy: 0.9994 - val_loss: 0.0070
Epoch 2/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 13s 3ms/step - accuracy: 0.9990 - loss: 0.0063 - val_accuracy: 0.9994 - val_loss: 0.0076
Epoch 3/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 13s 3ms/step - accuracy: 0.9990 - loss: 0.0058 - val_accuracy: 0.9994 - val_loss: 0.0068
Epoch 4/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 23s 4ms/step - accuracy: 0.9991 - loss: 0.0050 - val_accuracy: 0.9994 - val_loss: 0.0057
Epoch 5/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 17s 3ms/step - accuracy: 0.9992 - loss: 0.0043 - val_accuracy: 0.9994 - val_loss: 0.0064
Epoch 6/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 9s 3ms/step - accuracy: 0.9992 - loss: 0.0038 - val_accuracy: 0.9994 - val_loss: 0.0068
Epoch 7/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.9992 - loss: 0.0040 - val_accuracy: 0.9994 - val_loss: 0.0067
Epoch 8/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.9993 - loss: 0.

# Modelo 2 - Arquitetura com diferentes camadas e quantidades de neurônios (Batch).


In [14]:
model_batchnorm = Sequential([
    Dense(128, input_dim=X_train.shape[1]),
    BatchNormalization(),
    tf.keras.layers.Activation('relu'),
    Dropout(0.5),  # Adicionando Dropout para evitar overfitting
    Dense(64),
    BatchNormalization(),
    tf.keras.layers.Activation('relu'),
    Dropout(0.5),
    Dense(1, activation='sigmoid')  # Como é um problema binário, usamos sigmoid
])

model_batchnorm.compile(optimizer='adam',
                        loss='binary_crossentropy',
                        metrics=['accuracy'])

history_batchnorm = model_batchnorm.fit(X_train, y_train,
                                        epochs=10,
                                        batch_size=64,
                                        validation_data=(X_test, y_test))

Epoch 1/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 19s 4ms/step - accuracy: 0.9773 - loss: 0.0719 - val_accuracy: 0.9994 - val_loss: 0.0041
Epoch 2/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 12s 4ms/step - accuracy: 0.9992 - loss: 0.0048 - val_accuracy: 0.9994 - val_loss: 0.0043
Epoch 3/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.9992 - loss: 0.0048 - val_accuracy: 0.9994 - val_loss: 0.0038
Epoch 4/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.9993 - loss: 0.0040 - val_accuracy: 0.9994 - val_loss: 0.0040
Epoch 5/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 12s 4ms/step - accuracy: 0.9992 - loss: 0.0039 - val_accuracy: 0.9994 - val_loss: 0.0049
Epoch 6/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 17s 3ms/step - accuracy: 0.9993 - loss: 0.0038 - val_accuracy: 0.9994 - val_loss: 0.0041
Epoch 7/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 13s 4ms/step - accuracy: 0.9993 - loss: 0.0043 - val_accuracy: 0.9994 - val_loss: 0.0035
Epoch 8/10
3116/3116 ━━━━━━━━━━━━━━━━━━━━ 12s 4ms/step - accuracy: 0.9992 - loss: 0

# Calculando a Acurácia

In [15]:
test_loss1, test_accuracy1 = model.evaluate(X_test, y_test)
test_loss2, test_accuracy2 = model_batchnorm.evaluate(X_test, y_test)

print(f"Acurácia do Modelo 1: {test_accuracy1}")
print(f"Acurácia do Modelo 2: {test_accuracy2}")

2671/2671 ━━━━━━━━━━━━━━━━━━━━ 6s 2ms/step - accuracy: 0.9993 - loss: 0.0065
2671/2671 ━━━━━━━━━━━━━━━━━━━━ 3s 1ms/step - accuracy: 0.9993 - loss: 0.0042
Acurácia do Modelo 1: 0.9993680119514465
Acurácia do Modelo 2: 0.9993680119514465


# Cálculo do AUC-ROC

In [16]:
y_pred1 = (model.predict(X_test) > 0.5).astype("int32")
y_pred2 = (model_batchnorm.predict(X_test) > 0.5).astype("int32")

# Calculando AUC-ROC para ambos os modelos
auc_roc1 = roc_auc_score(y_test, y_pred1)
auc_roc2 = roc_auc_score(y_test, y_pred2)

print(f"AUC-ROC do Modelo 1: {auc_roc1}")
print(f"AUC-ROC do Modelo 2: {auc_roc2}")

2671/2671 ━━━━━━━━━━━━━━━━━━━━ 7s 3ms/step
2671/2671 ━━━━━━━━━━━━━━━━━━━━ 3s 1ms/step
AUC-ROC do Modelo 1: 0.9189300891796344
AUC-ROC do Modelo 2: 0.9152594797751237


# Criação do relatório para compararmos ambos modelos

In [19]:
print("Modelo 1")
print(classification_report(y_test, y_pred1))

print("Modelo 2 ")
print(classification_report(y_test, y_pred2))

Modelo 1
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85307
           1       0.78      0.84      0.81       136

    accuracy                           1.00     85443
   macro avg       0.89      0.92      0.90     85443
weighted avg       1.00      1.00      1.00     85443

Modelo 2 
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     85307
           1       0.78      0.83      0.81       136

    accuracy                           1.00     85443
   macro avg       0.89      0.92      0.90     85443
weighted avg       1.00      1.00      1.00     85443



# Análises e conclusões

**Acurácia de treinamento:** A acurácia dos modelos durante o treinamento e validação parece muito alta, próxima de 100%. ALgo bem comum e esperado nesses problemas de detecção de fraudes em cartões de crédito. No entanto, uma acurácia alta não necessariamente indica um bom desempenho na detecção de fraudes, que é a classe minoritária e a mais relevante neste contexto.

Outro fator de grande relevância a ser analisado é **AUC-ROC** O AUC-ROC que é uma métrica importante para avaliar a capacidade do modelo em diferenciar entre as classes  (fraude) e (não fraude). Sabemos que essa métrica varia entre 0-1, e em ambos os modelos os valores de AUC_ROC foram bem próximos. O valor do modelo 1 (0.9189) foi maior do que o valor encontrado no modelo 2(0.9153) que indica que ele tem uma capacidade ligeiramente melhor de distinguir entre fraudes e não fraudes em comparação com o Modelo 2. Porém vemos que a diferença entre os dois valores é pequena, sugerindo que ambos os modelos têm desempenho  similar.

Os **relatório de classificação** fornecem uma visão de três métricas principais: precision, recall e f1-score.

A precisão para a classe de fraude (classe 1) é de 0.78 para ambos os modelos, o que mostra que o modelo prevê uma fraude, ele está correto 78% das vezes.

O recall é maior no Modelo 1 (0.84) em comparação ao Modelo 2 (0.83). Esse valor indica que o Modelo 1 consegue identificar uma maior quantidade das fraudes presentes no conjunto de dados.

Ambos os modelos possuem um f1-score de 0.81, indicando que o equilíbrio entre precisão e recall é semelhante em ambos os casos.

**Conclusão**:
Ambos os modelos têm desempenho muito semelhante em termos de precisão, recall, f1-score e AUC-ROC. No entanto, o Modelo 1 se destaca com um AUC-ROC um pouco maior. Dessa forma, o modelo 1 pode ser preferido por apresentar valores maiores em algumas métricas o que o torna um pouco mais eficaz na detecção de fraudes sem comprometer muito a precisão. O modelo 2, apesar de ter um AUC-ROCinferior ainda apresenta um desempenho bom e também poderia ser escolhido.









